File: 2024-06/dna_palindrome.js
// 2020113486 김대건

const fs = require("fs");
const path = require("path");

function findPalindromes(sequence, minLen = 20, maxLen = 25, bufferLen = 10) {
    const palindromes = [];
    for (let i = 0; i < sequence.length - (22 + bufferLen + maxLen); i++) {
        const seq1 = sequence.slice(i, i + 22);
        const buffer = sequence.slice(i + 22, i + 32);
        for (let j = 20; j <= 25; j++) {
            const seq2 = sequence
                .slice(i + 32, i + 32 + j)
                .split("")
                .reverse()
                .join("");
            palindromes.push({ seq1, seq2, buffer, pos: i });
        }
    }
    return palindromes;
}

function localAlignment(seq1, seq2, match = 1, mismatch = -1, gap = -1) {
    const m = seq1.length;
    const n = seq2.length;
    const scoreMatrix = Array.from({ length: m + 1 }, () => Array(n + 1).fill(0));
    const backtrackMatrix = Array.from({ length: m + 1 }, () => Array(n + 1).fill(0));

    let maxScore = 0;
    let alignments = [];

    for (let i = 1; i <= m; i++) {
        for (let j = 1; j <= n; j++) {
            const matchScore = scoreMatrix[i - 1][j - 1] + (seq1[i - 1] === seq2[j - 1] ? match : mismatch);
            const deleteScore = scoreMatrix[i - 1][j] + gap;
            const insertScore = scoreMatrix[i][j - 1] + gap;
            scoreMatrix[i][j] = Math.max(0, matchScore, deleteScore, insertScore);

            if (scoreMatrix[i][j] === matchScore) {
                backtrackMatrix[i][j] = 1;
            } else if (scoreMatrix[i][j] === deleteScore) {
                backtrackMatrix[i][j] = 2;
            } else if (scoreMatrix[i][j] === insertScore) {
                backtrackMatrix[i][j] = 3;
            }

            if (scoreMatrix[i][j] > maxScore) {
                maxScore = scoreMatrix[i][j];
            }
            if (scoreMatrix[i][j] >= 18) {
                alignments.push({ i, j, score: scoreMatrix[i][j] });
            }
        }
    }

    const results = alignments.map(({ i, j, score }) => {
        let alignedSeq1 = "";
        let alignedSeq2 = "";
        while (scoreMatrix[i][j] !== 0) {
            if (backtrackMatrix[i][j] === 1) {
                alignedSeq1 = seq1[i - 1] + alignedSeq1;
                alignedSeq2 = seq2[j - 1] + alignedSeq2;
                i--;
                j--;
            } else if (backtrackMatrix[i][j] === 2) {
                alignedSeq1 = seq1[i - 1] + alignedSeq1;
                alignedSeq2 = "-" + alignedSeq2;
                i--;
            } else if (backtrackMatrix[i][j] === 3) {
                alignedSeq1 = "-" + alignedSeq1;
                alignedSeq2 = seq2[j - 1] + alignedSeq2;
                j--;
            }
        }
        return { score, alignedSeq1, alignedSeq2 };
    });

    return results;
}

if (process.argv.length !== 4) {
    console.log("Usage: node <program name> <inputfile> <outputfile>");
    process.exit(1);
}

const inputFile = path.resolve(process.argv[2]);
const outputFile = path.resolve(process.argv[3]);

fs.readFile(inputFile, "utf8", (err, data) => {
    if (err) {
        console.error(err);
        process.exit(1);
    }

    const sequence = data.trim();
    const palindromes = findPalindromes(sequence);

    let results = [];
    palindromes.forEach(({ seq1, seq2, buffer, pos }) => {
        const alignmentResults = localAlignment(seq1, seq2);
        alignmentResults.forEach((result) => {
            if (result.score >= 18) {
                results.push({ pos, buffer, ...result });
            }
        });
    });

    const output = results
        .map(
            ({ pos, buffer, score, alignedSeq1, alignedSeq2 }) =>
                `------------------------------\n` + `Position in file: ${pos}\n` + `Buffer: ${buffer}\n` + `Seq 1: ${alignedSeq1}\n` + `Seq 2: ${alignedSeq2}\n` + `Alignment score: ${score}\n`
        )
        .join("\n");

    fs.writeFile(outputFile, output, "utf8", (err) => {
        if (err) {
            console.error(err);
            process.exit(1);
        }
        console.log("Results written to", outputFile);
    });
});

File: 2024-06/dna_palindrome.py
import sys

def find_palindromes(sequence, min_len=20, max_len=25, buffer_len=10):
    palindromes = []
    for i in range(len(sequence) - (22 + buffer_len + max_len)):
        seq1 = sequence[i:i+22]
        buffer = sequence[i+22:i+32]
        for j in range(20, 26):
            seq2 = sequence[i+32:i+32+j][::-1]
            palindromes.append((seq1, seq2, i))
    return palindromes

def local_alignment(seq1, seq2, match=1, mismatch=-1, gap=-1):
    m, n = len(seq1), len(seq2)
    score_matrix = [[0 for _ in range(n+1)] for _ in range(m+1)]
    backtrack_matrix = [[0 for _ in range(n+1)] for _ in range(m+1)]
    
    max_score = 0
    max_pos = None
    alignments = []

    for i in range(1, m+1):
        for j in range(1, n+1):
            match_score = score_matrix[i-1][j-1] + (match if seq1[i-1] == seq2[j-1] else mismatch)
            delete = score_matrix[i-1][j] + gap
            insert = score_matrix[i][j-1] + gap
            score_matrix[i][j] = max(0, match_score, delete, insert)

            if score_matrix[i][j] == match_score:
                backtrack_matrix[i][j] = 1
            elif score_matrix[i][j] == delete:
                backtrack_matrix[i][j] = 2
            elif score_matrix[i][j] == insert:
                backtrack_matrix[i][j] = 3

            if score_matrix[i][j] > max_score:
                max_score = score_matrix[i][j]
                max_pos = (i, j)
            if score_matrix[i][j] >= 18:
                alignments.append((i, j, score_matrix[i][j]))

    results = []
    for i, j, score in alignments:
        aligned_seq1, aligned_seq2 = '', ''
        while score_matrix[i][j] != 0:
            if backtrack_matrix[i][j] == 1:
                aligned_seq1 = seq1[i-1] + aligned_seq1
                aligned_seq2 = seq2[j-1] + aligned_seq2
                i -= 1
                j -= 1
            elif backtrack_matrix[i][j] == 2:
                aligned_seq1 = seq1[i-1] + aligned_seq1
                aligned_seq2 = '-' + aligned_seq2
                i -= 1
            elif backtrack_matrix[i][j] == 3:
                aligned_seq1 = '-' + aligned_seq1
                aligned_seq2 = seq2[j-1] + aligned_seq2
                j -= 1
        results.append((max_pos, score, aligned_seq1, aligned_seq2))

    return results

if __name__ == "__main__":
    input_file = sys.argv[1]
    
    with open(input_file, 'r') as file:
        sequence = file.read().strip()
    
    palindromes = find_palindromes(sequence)
    
    results = []
    for seq1, seq2, pos in palindromes:
        alignment_results = local_alignment(seq1, seq2)
        for result in alignment_results:
            max_pos, score, aligned_seq1, aligned_seq2 = result
            if score >= 18:
                results.append((pos, score, aligned_seq1, aligned_seq2))
    
    with open("output.txt", 'w') as file:
        for pos, score, aligned_seq1, aligned_seq2 in results:
            file.write("------------------------------\n")
            file.write(f"Position in file: {pos}\n")
            file.write(f"Seq 1: {aligned_seq1}\n")
            file.write(f"Seq 2: {aligned_seq2}\n")
            file.write(f"Alignment score: {score}\n")
            file.write("\n")
